### ISSUE TITLE
no type inferences for constant assignments of literal arrays

### ISSUE THREAD

**Issue by lgo** [sorbet/sorbet#1422]
Related to #1421, where constant inference is not handled. I kept this issue separate because the ergonomics about inferring percent literal arrays may be trickier.

From earlier conversation with @elliottt, using `%w{...}` as an example, these are intuitively `T::Array[String]` but a more strict and correct type would be `[String, String, ...]`. The stricter type is required for any literals that are splatted, as `T::Array` cannot be splatted.

The ergonomics around these are tricky because any time the literal is modified, you'd have to go and replace the type definition of it. If any generated type sigs also used `[String, String, ...]` instead, they'd need to be updated to, but in most practical cases those types can just be be `T::Array[String]`. 

#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0A%0Aextend%20T%3A%3ASig%0A%0A%23%20All%20of%20these%20constant%20assignments%20are%20just%20T.untyped.%0A%0Amodule%20Foo%0A%20%20PercentStringArrayLiteral%20%3D%20%25w%7Bhello%7D%0A%20%20StringArrayLiteral%20%3D%20%5B%22hello%22%5D%0A%20%20PercentSymbolArrayLiteral%20%3D%20%25i%7Bhello%7D%0Aend%0A%0A%23%20These%20are%20tricky%20because%20these%20types%20are%20more%20correct%20as%0A%23%20%5BString%2C%20String%2C%20String%2C%20String%2C%20...%5D%0A%23%20But%20using%20these%20types%20is%20not%20very%20ergonomical.%20Everytime%20a%0A%23%20new%20value%20is%20added%2C%20the%20type%20and%20all%20sigs%20with%20the%20type%20need%0A%23%20to%20be%20updated.%0AT.reveal_type(Foo%3A%3APercentStringArrayLiteral)%0AT.reveal_type(Foo%3A%3APercentSymbolArrayLiteral)%0AT.reveal_type(Foo%3A%3AStringArrayLiteral)%0A%0Astring_array_literal%20%3D%20%25w%7Bhello%7D%0Asymbol_array_literal%20%3D%20%25i%7Bhello%7D%0AT.reveal_type(string_array_literal)%0AT.reveal_type(symbol_array_literal)%0A)

```ruby
# typed: true

extend T::Sig

# All of these constant assignments are just T.untyped.

module Foo
  PercentStringArrayLiteral = %w{hello}
  StringArrayLiteral = ["hello"]
  PercentSymbolArrayLiteral = %i{hello}
end

# These are tricky because these types are more correct as
# [String, String, String, String, ...]
# But using these types is not very ergonomical. Everytime a
# new value is added, the type and all sigs with the type need
# to be updated.
T.reveal_type(Foo::PercentStringArrayLiteral)
T.reveal_type(Foo::PercentSymbolArrayLiteral)
T.reveal_type(Foo::StringArrayLiteral)

# These have the correct types.
string_array_literal = %w{hello}
symbol_array_literal = %i{hello}
T.reveal_type(string_array_literal)
T.reveal_type(symbol_array_literal)

```

#### Observed output
 
Constant assignments are `T.untyped`.

#### Expected behavior

Constant assignments will be their respective types, rather than `T.untyped`.

**Comment by jez** [comment]
This is basically a dupe of #1421 in the sense that if we solve this, it should be via a mechanism that also solves that, rather than hard-coding either array literals or Regexp literals. See https://github.com/sorbet/sorbet/issues/1421#issuecomment-516100985

**Comment by jez** [comment]
Fixed in #7993
